### ISSUE TITLE
[Crash]: OpenTTD crashed on MacOS with 32bpp baseset due to touchbar

### ISSUE THREAD

**Issue by ryanhsu999** [OpenTTD/OpenTTD#9876]
### Version of OpenTTD

OpenTTD 12.2 on MacOS Monterey 12.3.1

### Steps to reproduce

N/A

### Upload crash files

[crash20220503214114.log](https://github.com/OpenTTD/OpenTTD/files/8616920/crash20220503214114.log)
![crash20220503214114](https://user-images.githubusercontent.com/104873246/166594468-62dbf127-1408-4fcb-8074-76febfb9a9df.png)

**Comment by James103** [comment]
Do you have a `crash20220503214114.dmp` or a recent `crash.dmp`? If so, can you please zip the file and post it here?

**Comment by ryanhsu999** [comment]
Sorry I only saw log, png, and sav files. Github does not allow sav file
upload so I omitted it. It is attached to this email.

**Comment by James103** [comment]
> Github does not allow sav file upload so I omitted it. It is attached to this email

Any attachments sent via an email to GitHub are not copied over to the resulting comment. Will you please zip `crash20220503214114.sav` or a recent `crash.sav` and upload it here?

**Comment by Ryan9000** [comment]
[crash20220503214114.sav.zip](https://github.com/OpenTTD/OpenTTD/files/8637460/crash20220503214114.sav.zip)

**Comment by JGRennison** [comment]
It is probably worth trying with a non-32bpp baseset (i.e. OpenGFX) instead of using abase/zbase

**Comment by adin3d** [comment]
same problem for me, can not start it on Mac


`*** OpenTTD Crash Report ***

Crash at: 2022-05-26 13:31:21 (UTC)
In game date: 2185-12-02 (40)

Crash reason:
 Signal:  Segmentation fault: 11 (11)
 Message: <none>

OpenTTD version:
 Version:    12.2 (0)
 NewGRF ver: 1c286d64
 Bits:       64
 Endian:     little
 Dedicated:  no
 Build date: Apr  2 2022 10:56:53


Stacktrace:
 [00] openttd              0x00000001049719f3 (Pool<Company, Owner, 1ul, 15ul, (PoolType)1, false, true>::CleanPool() + 226451)
 [01] openttd              0x000000010482dcc3 (Pool<ServerNetworkGameSocketHandler, unsigned char, 8ul, 256ul, (PoolType)2, false, true>::CleanPool() + 70851)
 [02] openttd              0x000000010482db92 (Pool<ServerNetworkGameSocketHandler, unsigned char, 8ul, 256ul, (PoolType)2, false, true>::CleanPool() + 70546)
 [03] libsystem_platform.dylib 0x00007ff80e1a3dfd (_sigtramp + 29)
 [04] openttd              0x000000010678eba0
 [05] openttd              0x00000001049c9316 (SaveLoadWindow::SaveHeightmapConfirmationCallback(Window*, bool) + 100326)
 [06] openttd              0x00000001049c8de4 (SaveLoadWindow::SaveHeightmapConfirmationCallback(Window*, bool) + 98996)
 [07] openttd              0x00000001048f6fef (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 287967)
 [08] openttd              0x00000001048f2d4b (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 270907)
 [09] openttd              0x0000000104905502 (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 346610)
 [10] openttd              0x00000001048f3cd1 (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 274881)
 [11] openttd              0x00000001048f5c21 (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 282897)
 [12] CoreFoundation       0x00007ff80e24a03c (__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12)
 [13] CoreFoundation       0x00007ff80e2e7592 (___CFXRegistrationPost_block_invoke + 49)
 [14] CoreFoundation       0x00007ff80e2e7510 (_CFXRegistrationPost + 496)
 [15] CoreFoundation       0x00007ff80e21bca8 (_CFXNotificationPost + 735)
 [16] Foundation           0x00007ff80f05090e (-[NSNotificationCenter postNotificationName:object:userInfo:] + 82)
 [17] CoreFoundation       0x00007ff80e24a03c (__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12)
 [18] CoreFoundation       0x00007ff80e2e7592 (___CFXRegistrationPost_block_invoke + 49)
 [19] CoreFoundation       0x00007ff80e2e7510 (_CFXRegistrationPost + 496)
 [20] CoreFoundation       0x00007ff80e21bca8 (_CFXNotificationPost + 735)
 [21] Foundation           0x00007ff80f05090e (-[NSNotificationCenter postNotificationName:object:userInfo:] + 82)
 [22] AppKit               0x00007ff810c81dee (-[NSApplication _postDidFinishNotification] + 305)
 [23] AppKit               0x00007ff810c81b3c (-[NSApplication _sendFinishLaunchingNotification] + 208)
 [24] AppKit               0x00007ff810c7f70c (-[NSApplication(NSAppleEventHandling) _handleAEOpenEvent:] + 541)
 [25] AppKit               0x00007ff810c7f360 (-[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 665)
 [26] Foundation           0x00007ff80f07b834 (-[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 308)
 [27] Foundation           0x00007ff80f07b6a6 (_NSAppleEventManagerGenericHandler + 80)
 [28] AE                   0x00007ff8148ec3f0 (_AppleEventsCheckInAppWithBlock + 14079)
 [29] AE                   0x00007ff8148ebc5a (_AppleEventsCheckInAppWithBlock + 12137)
 [30] AE                   0x00007ff8148e538f (aeProcessAppleEvent + 419)
 [31] HIToolbox            0x00007ff816eeb3d2 (AEProcessAppleEvent + 54)
 [32] AppKit               0x00007ff810c79992 (_DPSNextEvent + 2036)
 [33] AppKit               0x00007ff810c77bfa (-[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1394)
 [34] AppKit               0x00007ff810c6a2a9 (-[NSApplication run] + 586)
 [35] openttd              0x0000000104aebb46 (BuildObjectWindow::TagNameFilter(ObjectClassID const*, StringFilter&) + 25670)
 [36] openttd              0x0000000104833412 (Pool<ServerNetworkGameSocketHandler, unsigned char, 8ul, 256ul, (PoolType)2, false, true>::CleanPool() + 93202)
 [37] ???                  0x000000010efac51e

Operating system:
 Name:     Mac OS X
 Release:  12.3.1
 Machine:  Intel x86-64h Haswell
 Min Ver:  101400
 Max Ver:  110100
 Compiler: GCC 4.2.1 "Apple LLVM 12.0.0 (clang-1200.0.32.29)"

Configuration:
 Blitter:      32bpp-sse4-anim
 Graphics set: zBase (5588)
 Language:     /Users/jamesr/Library/Application Support/Steam/steamapps/common/OpenTTD/OpenTTD.app/Contents/Resources/lang/english_AU.lng
 Music driver: cocoa
 Music set:    OpenMSX (7761)
 Network:      no
 Sound driver: cocoa
 Sound set:    NoSound (2)
 Video driver: cocoa

Fonts:
 Small:  sprite
 Medium: sprite
 Large:  sprite
 Mono:   sprite

AI Configuration (local: 0) (current: 0):
  0: Human

Libraries:
 LZMA:       5.2.5
 LZO:        2.10
 PNG:        1.6.37
 Zlib:       1.2.11

---- gamelog start ----
Tick 0: new game started
Tick 1280: game loaded
Tick 1280: game loaded
Tick 1280: game loaded
Tick 10478: game loaded
Tick 13584: game loaded
Tick 47111: game loaded
Tick 51965: game loaded
Tick 62555: game loaded
Tick 63600: game loaded
Tick 49396: game loaded
Tick 62604: game loaded
Tick 64272: game loaded
Tick 6006: game loaded
Tick 17767: game loaded
Tick 49394: game loaded
Tick 30847: game loaded
Tick 60372: game loaded
Revision text changed to g79dc634d41d04, savegame version 297, not modified, _openttd_newgrf_version = 0x1c006d64
New game mode: 1 landscape: 1
Tick 62318: game loaded
Revision text changed to 12.2, savegame version 299, not modified, _openttd_newgrf_version = 0x1c286d64
New game mode: 0 landscape: 1
---- gamelog end ----

Recent news messages:

*** End of OpenTTD Crash Report ***
`

**Comment by Nesvet** [comment]
Same here. Just switching "Base graphics set" in "Game Options" to anything else default "OpenGFX" - and Boom! Crash

**Comment by James103** [comment]
> Same here. Just switching "Base graphics set" in "Game Options" to anything else default "OpenGFX" - and Boom! Crash

Do you have the crash log and/or crash dump? If so, can you please upload them here?

**Comment by Nesvet** [comment]
> > Same here. Just switching "Base graphics set" in "Game Options" to anything else default "OpenGFX" - and Boom! Crash
> 
> Do you have the crash log and/or crash dump? If so, can you please upload them here?

A lot. There is the log one ( [crash20220608065832.log](https://github.com/OpenTTD/OpenTTD/files/8864407/crash20220608065832.log) ). If you want png or sav one, let me know

**Comment by nielsmh** [comment]
So far all the reports for this issue use the English (AU) language option. It's unlikely, but can you try changing language in the game to English (UK) instead and see if the crash still happens?

**Comment by JGRennison** [comment]
Probably this is the same as https://github.com/OpenTTD/OpenTTD/issues/9788 which is caused by touchbar support

**Comment by Nesvet** [comment]
> So far all the reports for this issue use the English (AU) language option. It's unlikely, but can you try changing language in the game to English (UK) instead and see if the crash still happens?

Yes, now it works! 👍

**Comment by nielsmh** [comment]
So there is something wrong with the Australian language file, and it causes the game to crash. That's definitely a problem.

**Comment by James103** [comment]
On what version, nightly, or commit did the crash start to occur?

**Comment by JakubR01** [comment]
It isn't only problem on Mac and with that language. I have crashes on Linux machine and with polish language. Dutch crashes too after fast change.

**Comment by Eilon** [comment]
I'm also getting a crash, probably the same one. I installed various 32bpp NewGRFs from https://www.reddit.com/r/openttd/comments/wghmgy/real_industries_32bpp_beta_released_download_at/, and that was fine. Then I downloaded zBase, and when I switched to select that in the game settings, it immediately crashed. Next time I open OpenTTD it went back to OpenGFX, and everything is fine, until I try to switch to zBase again, and I get the same crash.

This is a clean install of OpenTTD 12.2 on macOS (Intel, TouchBar).

* [crash20220810060534.log](https://github.com/OpenTTD/OpenTTD/files/9297232/crash20220810060534.log)
* [crash20220810060534.sav.zip](https://github.com/OpenTTD/OpenTTD/files/9297237/crash20220810060534.sav.zip)

![crash20220810060534](https://user-images.githubusercontent.com/202643/183827959-6e173cf3-4e04-4e2b-9f2b-3ca3931a8e90.png)

**Comment by beyre83** [comment]
the crash is related to the Touch Bar on the MacBook Pro's it is present in all versions from 12.0 - 12.2 inc latest nightly's when using either abase/zbase id assume its something todo with rendering of the Touch Bar icons for the game... however 12.0 RC1 works perfectly with the Touch Bar and abase/zbase with out the aforementioned crashes that the above posters have posted so maybe worth a look at the code between the 12.0 rc1 and 12.x releases since to see if there's some small change that is causing the crash hopefully this information will help someone be able to fix this issue

**Comment by Bouke** [comment]
I've tried replicating this issue, but it works fine on my end. This is on an Intel MacBook with TouchBar running macOS Monterey (12.5.1). OpenTTD 12.2, tested with languages Dutch, Polish, English (AU) and both zBase 2015-04-20 and aBase 2016-05-17. Any suggestions for reproducing this crash?

**Comment by beyre83** [comment]
I am downloading 12.5.1 Monterey to see if this has resolved the issue as it is present in 12.5 Monterey on intel MacBook Pro with the zbase & ottd 12.x.. will update if the os update resolves the issue would be good if someone on 12.6 Monterey could confirm if OTTD 12.X works as intended as well

**Comment by adin3d** [comment]
still crashes for me :(

from the moment I try to open it from Steam, so I don't have access to any settings 

*** OpenTTD Crash Report ***

Crash at: 2022-09-24 16:55:55 (UTC)
In game date: 2185-12-02 (40)

Crash reason:
 Signal:  Segmentation fault: 11 (11)
 Message: <none>

OpenTTD version:
 Version:    12.2 (0)
 NewGRF ver: 1c286d64
 Bits:       64
 Endian:     little
 Dedicated:  no
 Build date: Apr  2 2022 10:56:53


Stacktrace:
 [00] openttd              0x000000010cda69f3 (Pool<Company, Owner, 1ul, 15ul, (PoolType)1, false, true>::CleanPool() + 226451)
 [01] openttd              0x000000010cc62cc3 (Pool<ServerNetworkGameSocketHandler, unsigned char, 8ul, 256ul, (PoolType)2, false, true>::CleanPool() + 70851)
 [02] openttd              0x000000010cc62b92 (Pool<ServerNetworkGameSocketHandler, unsigned char, 8ul, 256ul, (PoolType)2, false, true>::CleanPool() + 70546)
 [03] libsystem_platform.dylib 0x00007ff8072ecdfd (_sigtramp + 29)
 [04] openttd              0x000000010ebc3ba0
 [05] openttd              0x000000010cdfe316 (SaveLoadWindow::SaveHeightmapConfirmationCallback(Window*, bool) + 100326)
 [06] openttd              0x000000010cdfdde4 (SaveLoadWindow::SaveHeightmapConfirmationCallback(Window*, bool) + 98996)
 [07] openttd              0x000000010cd2bfef (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 287967)
 [08] openttd              0x000000010cd27d4b (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 270907)
 [09] openttd              0x000000010cd3a502 (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 346610)
 [10] openttd              0x000000010cd28cd1 (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 274881)
 [11] openttd              0x000000010cd2ac21 (SaveFilter* CreateSaveFilter<LZMASaveFilter>(SaveFilter*, unsigned char) + 282897)
 [12] CoreFoundation       0x00007ff80739303c (__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12)
 [13] CoreFoundation       0x00007ff807430592 (___CFXRegistrationPost_block_invoke + 49)
 [14] CoreFoundation       0x00007ff807430510 (_CFXRegistrationPost + 496)
 [15] CoreFoundation       0x00007ff807364ca8 (_CFXNotificationPost + 735)
 [16] Foundation           0x00007ff80819990e (-[NSNotificationCenter postNotificationName:object:userInfo:] + 82)
 [17] CoreFoundation       0x00007ff80739303c (__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12)
 [18] CoreFoundation       0x00007ff807430592 (___CFXRegistrationPost_block_invoke + 49)
 [19] CoreFoundation       0x00007ff807430510 (_CFXRegistrationPost + 496)
 [20] CoreFoundation       0x00007ff807364ca8 (_CFXNotificationPost + 735)
 [21] Foundation           0x00007ff80819990e (-[NSNotificationCenter postNotificationName:object:userInfo:] + 82)
 [22] AppKit               0x00007ff809dcadee (-[NSApplication _postDidFinishNotification] + 305)
 [23] AppKit               0x00007ff809dcab3c (-[NSApplication _sendFinishLaunchingNotification] + 208)
 [24] AppKit               0x00007ff809dc870c (-[NSApplication(NSAppleEventHandling) _handleAEOpenEvent:] + 541)
 [25] AppKit               0x00007ff809dc8360 (-[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 665)
 [26] Foundation           0x00007ff8081c4834 (-[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 308)
 [27] Foundation           0x00007ff8081c46a6 (_NSAppleEventManagerGenericHandler + 80)
 [28] AE                   0x00007ff80da353f0 (_AppleEventsCheckInAppWithBlock + 14079)
 [29] AE                   0x00007ff80da34c5a (_AppleEventsCheckInAppWithBlock + 12137)
 [30] AE                   0x00007ff80da2e38f (aeProcessAppleEvent + 419)
 [31] HIToolbox            0x00007ff8100343d2 (AEProcessAppleEvent + 54)
 [32] AppKit               0x00007ff809dc2992 (_DPSNextEvent + 2036)
 [33] AppKit               0x00007ff809dc0bfa (-[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1394)
 [34] AppKit               0x00007ff809db32a9 (-[NSApplication run] + 586)
 [35] openttd              0x000000010cf20b46 (BuildObjectWindow::TagNameFilter(ObjectClassID const*, StringFilter&) + 25670)
 [36] openttd              0x000000010cc68412 (Pool<ServerNetworkGameSocketHandler, unsigned char, 8ul, 256ul, (PoolType)2, false, true>::CleanPool() + 93202)
 [37] ???                  0x000000011e9be51e

Operating system:
 Name:     Mac OS X
 Release:  12.3.1
 Machine:  Intel x86-64h Haswell
 Min Ver:  101400
 Max Ver:  110100
 Compiler: GCC 4.2.1 "Apple LLVM 12.0.0 (clang-1200.0.32.29)"

Configuration:
 Blitter:      32bpp-sse4-anim
 Graphics set: zBase (5588)
 Language:     /Users/george/Library/Application Support/Steam/steamapps/common/OpenTTD/OpenTTD.app/Contents/Resources/lang/english_AU.lng
 Music driver: cocoa
 Music set:    OpenMSX (7761)
 Network:      no
 Sound driver: cocoa
 Sound set:    NoSound (2)
 Video driver: cocoa

Fonts:
 Small:  sprite
 Medium: sprite
 Large:  sprite
 Mono:   sprite

AI Configuration (local: 0) (current: 0):
  0: Human

Libraries:
 LZMA:       5.2.5
 LZO:        2.10
 PNG:        1.6.37
 Zlib:       1.2.11

---- gamelog start ----
Tick 0: new game started
Tick 1280: game loaded
Tick 1280: game loaded
Tick 1280: game loaded
Tick 10478: game loaded
Tick 13584: game loaded
Tick 47111: game loaded
Tick 51965: game loaded
Tick 62555: game loaded
Tick 63600: game loaded
Tick 49396: game loaded
Tick 62604: game loaded
Tick 64272: game loaded
Tick 6006: game loaded
Tick 17767: game loaded
Tick 49394: game loaded
Tick 30847: game loaded
Tick 60372: game loaded
Revision text changed to g79dc634d41d04, savegame version 297, not modified, _openttd_newgrf_version = 0x1c006d64
New game mode: 1 landscape: 1
Tick 62318: game loaded
Revision text changed to 12.2, savegame version 299, not modified, _openttd_newgrf_version = 0x1c286d64
New game mode: 0 landscape: 1
---- gamelog end ----

Recent news messages:

*** End of OpenTTD Crash Report ***

**Comment by Bouke** [comment]
After various attempts I have managed to reproduce this issue. I have found that to crash:

- hardware acceleration must be off; 32bpp_anim_sse4 blitter is used
- full screen must be off
- language doesn't seem to have an effect

This is on Monterey 12.5.1 on a MacBook with Touch Bar. The crash happens because the blitter is trying to read from what I think is a sprite's corrupted memory. The pause (726) sprite's data looks like this when running with the 40bpp blitter (OpenGL). Note the values of the first 48 bytes (offsets `uint32[6][2]`).

```
   00 00 00 00   c0 99 00 00   f0 ed 00 00   50 15 01 00   │ ············P··· │
   a8 2c 01 00   f8 36 01 00   f4 3d 01 00   c4 40 01 00   │ ·,···6···=···@·· │
   dc 42 01 00   dc 43 01 00   b8 44 01 00   08 45 01 00   │ ·B···C···D···E·· │
```

This is the same sprite's data for the 32bpp anim sse4 blitter:
```
   0e 00 00 00   00 00 00 00   c0 c6 00 00   a8 01 68 00   │ ··············h· │
   40 28 01 00   e0 5a 01 00   d8 00 34 00   40 73 01 00   │ @(···Z····4·@s·· │
   60 80 01 00   70 00 1a 00   78 86 01 00   fc 89 01 00   │ `···p···x······· │
```

I would expect the sprite to not have changed -- however that might be my lack of understanding how sprites or blitting works.

In any case, when running full screen the application doesn't crash, but the touch bar is missing most of the sprites:
<img width="1085" alt="image" src="https://user-images.githubusercontent.com/235882/192366414-8485bdf4-0999-441c-b0b0-20c8f7a8989f.png">

I assumed the fix would be as simple as the fix used for the 40bpp blitter, however that didn't work (probably because the sprite's data is corrupted?)
https://github.com/OpenTTD/OpenTTD/blob/e5203f34092903693276b76f5c09bf00530e025f/src/blitter/40bpp_anim.cpp#L44-L48

However the following simple patch disables rendering the touch bar sprites, with a similar condition to above. I hope it'll help debugging further into this issue:
```diff
diff --git a/src/blitter/32bpp_anim_sse4.cpp b/src/blitter/32bpp_anim_sse4.cpp
index 5089dce69..1249b431c 100644
--- a/src/blitter/32bpp_anim_sse4.cpp
+++ b/src/blitter/32bpp_anim_sse4.cpp
@@ -367,6 +367,12 @@ IGNORE_UNINITIALIZED_WARNING_STOP
  */
 void Blitter_32bppSSE4_Anim::Draw(Blitter::BlitterParams *bp, BlitterMode mode, ZoomLevel zoom)
 {
+	assert(_screen.dst_ptr != nullptr);
+
+	if (_screen_disable_anim) {
+		return;
+	}
+
 	const Blitter_32bppSSE_Base::SpriteFlags sprite_flags = ((const Blitter_32bppSSE_Base::SpriteData *) bp->sprite)->flags;
 	switch (mode) {
 		default: {
```

**Comment by 2TallTyler** [comment]
Fixed by #10060
